### PR TITLE
[Connectedk8s] Kubernetes distro and infra parameter and heuristics support

### PR DESCRIPTION
**Pull request by alphaWizard**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Connectedk8s

File: src/connectedk8s/azext_connectedk8s/_params.py [hunk 1]
```diff
@@ -5,10 +5,13 @@
...
 
 from argcomplete.completers import FilesCompleter
-from azure.cli.core.commands.parameters import get_location_type
+from azure.cli.core.commands.parameters import get_location_type, get_enum_type
 from azure.cli.core.commands.parameters import (file_type)
 from azure.cli.core.commands.validators import get_default_location_from_resource_group
 
+distribution_types = ["auto", "generic", "openshift", "rancher_RKE", "kind", "GKE", "EKS", "AKS", "AKS-HCI", "CAPZ", "AKS-engine", "TKG"]
+infrastructure_types = ["auto", "generic", "azure", "AWS", "GCP", "Azure_Stack_HC", "Azure_Stack_Hub", "Azure_Stack_Edge", "vSphere"]
+
```
Comment (alphaWizard) [inline src/connectedk8s/azext_connectedk8s/_params.py:14]: @shashank , "auto" is default cli, which triggers heuristics evaluation, Remaining are the enum values.

File: src/connectedk8s/azext_connectedk8s/_constants.py [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 # pylint: disable=line-too-long
 
+infrastructure_types = ["generic", "azure", "AWS", "GCP", "Azure_Stack_HC", "Azure_Stack_Hub", "Azure_Stack_Edge", "vSphere"]
```
Comment (shashankbarsin) [inline src/connectedk8s/azext_connectedk8s/_constants.py:9]: Have following values:

- generic
- AWS
- GCP
- Azure
- AzureStackEdge
- AzureStackHub
- AzureStackHCI
- vSphere

File: src/connectedk8s/azext_connectedk8s/_params.py [hunk 3]
```diff
@@ -5,10 +5,13 @@
 # pylint: disable=line-too-long
 
 from argcomplete.completers import FilesCompleter
-from azure.cli.core.commands.parameters import get_location_type
+from azure.cli.core.commands.parameters import get_location_type, get_enum_type
 from azure.cli.core.commands.parameters import (file_type)
 from azure.cli.core.commands.validators import get_default_location_from_resource_group
 
+distribution_types = ["auto", "generic", "openshift", "rancher_RKE", "kind", "GKE", "EKS", "AKS", "AKS-HCI", "CAPZ", "AKS-engine", "TKG"]
+infrastructure_types = ["auto", "generic", "azure", "AWS", "GCP", "Azure_Stack_HC", "Azure_Stack_Hub", "Azure_Stack_Edge", "vSphere"]
```
Comment (shashankbarsin) [inline src/connectedk8s/azext_connectedk8s/_params.py:13]: Use following values:

- generic
- AWS
- GCP
- Azure
- AzureStackEdge
- AzureStackHub
- AzureStackHCI
- vSphere

File: src/connectedk8s/azext_connectedk8s/_constants.py (hunk 2 above)
Comment (shashankbarsin) [reply src/connectedk8s/azext_connectedk8s/_constants.py:9]: vSphere I'll check if we can differentiate between 6.x and 7.x by heuristics

File: src/connectedk8s/azext_connectedk8s/_params.py [hunk 4]
```diff
@@ -5,10 +5,13 @@
 # pylint: disable=line-too-long
 
 from argcomplete.completers import FilesCompleter
-from azure.cli.core.commands.parameters import get_location_type
+from azure.cli.core.commands.parameters import get_location_type, get_enum_type
 from azure.cli.core.commands.parameters import (file_type)
 from azure.cli.core.commands.validators import get_default_location_from_resource_group
 
+distribution_types = ["auto", "generic", "openshift", "rancher_RKE", "kind", "GKE", "EKS", "AKS", "AKS-HCI", "CAPZ", "AKS-engine", "TKG"]
```
Comment (shashankbarsin) [inline src/connectedk8s/azext_connectedk8s/_params.py:12]: Use following values:

- Azure Kubernetes Service
- Google Kubernetes Engine
- Amazon Elastic Kubernetes Service
- Rancher Kubernetes Engine
- OpenShift Kubernetes Distribution
- OpenShift Container Platform
- K3s
- kind
- minikube
- Cluster API Provider Azure
- AKS Engine
- Tanzu Kubernetes Grid
- Canonical Charmed Kubernetes

Comment (shashankbarsin) [reply src/connectedk8s/azext_connectedk8s/_params.py:12]: We should check if we can distinguish between OKD and OCP in above list.

Comment (shashankbarsin) [reply src/connectedk8s/azext_connectedk8s/_params.py:12]: We should also check if Docker for [Mac](https://docs.docker.com/docker-for-mac/#kubernetes)/[Windows](https://docs.docker.com/docker-for-windows/#kubernetes) has any cluster identifying characteristic in which case we can include that in the list as well

File: src/connectedk8s/azext_connectedk8s/_params.py (hunk 3 above)
Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/_params.py:13]: One doubt, so if user is providing these values, he will have to type whole "Amazon_Web_Services". Isn't it easy for user for type just --infrastructure AWS?. We can always add a logic wherever we want to display(like in portal) to transform from AWS to Amazon Web Services.

**Comment by yonzhan** [comment]
Connectedk8s

File: src/connectedk8s/azext_connectedk8s/vendored_sdks/operations/connected_cluster_operations.py [hunk 5]
```diff
@@ -500,9 +515,7 @@ def internal_paging(next_link=None, raw=False):
             response = self._client.send(request, stream=False, **operation_config)
 
             if response.status_code not in [200]:
-                exp = CloudError(response)
-                exp.request_id = response.headers.get('x-ms-request-id')
-                raise exp
+                raise models.ErrorResponseException(self._deserialize, response)
```
Comment (akashkeshari) [inline src/connectedk8s/azext_connectedk8s/vendored_sdks/operations/connected_cluster_operations.py:518]: how is the error response experience in this case? Is it the same as before?

File: src/connectedk8s/azext_connectedk8s/_params.py [hunk 6]
```diff
@@ -5,10 +5,13 @@
 # pylint: disable=line-too-long
 
 from argcomplete.completers import FilesCompleter
-from azure.cli.core.commands.parameters import get_location_type
+from azure.cli.core.commands.parameters import get_location_type, get_enum_type
 from azure.cli.core.commands.parameters import (file_type)
 from azure.cli.core.commands.validators import get_default_location_from_resource_group
 
+distribution_types = ["auto", "generic", "openshift", "rancher_rke", "kind", "k3s", "minikube", "gke", "eks", "aks", "aks_hci", "capz", "aks_engine", "tkg"]
```
Comment (akashkeshari) [inline src/connectedk8s/azext_connectedk8s/_params.py:12]: Can you import the enum values from consts.py?
See Example: https://github.com/Azure/azure-cli-extensions/blob/master/src/aks-preview/azext_aks_preview/_params.py

File: src/connectedk8s/azext_connectedk8s/custom.py [hunk 7]
```diff
@@ -747,11 +803,9 @@ def update_agents(cmd, client, resource_group_name, cluster_name, https_proxy=""
 
     # Get kubernetes cluster info for telemetry
     kubernetes_version = get_server_version(configuration)
-    kubernetes_distro = get_kubernetes_distro(configuration)
 
     kubernetes_properties = {
-        'Context.Default.AzureCLI.KubernetesVersion': kubernetes_version,
-        'Context.Default.AzureCLI.KubernetesDistro': kubernetes_distro
```
Comment (akashkeshari) [inline src/connectedk8s/azext_connectedk8s/custom.py:754]: Why is this removed?

Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/custom.py:754]: Because, in update_agents we havent exposed parameters, and so, heuristics will set the distro and infra. But, if user has already provided the distro and infra through --distribution and --infrastructure, this will skew the helm parameters and telemetry with respect to the cc object. And we dont patch these distro/infra values in CC object now, as it was decided. kubernetes_version was removed by mistake I think, will replace it.

File: src/connectedk8s/azext_connectedk8s/_params.py (hunk 6 above)
Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/_params.py:12]: Okay, will update accordingly.

File: src/connectedk8s/azext_connectedk8s/vendored_sdks/operations/connected_cluster_operations.py (hunk 5 above)
Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/vendored_sdks/operations/connected_cluster_operations.py:518]: Will track these things, when we test in staging in next iteration.

File: src/connectedk8s/azext_connectedk8s/_params.py (hunk 6 above)
Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/_params.py:12]: I will add the two lists in constants.py and import those in _params.py.

File: src/connectedk8s/azext_connectedk8s/custom.py (hunk 7 above)
Comment (akashkeshari) [reply src/connectedk8s/azext_connectedk8s/custom.py:761]: Cant we do a get on the cc resource and use the distro value in telemetry?

Comment (alphaWizard) [reply src/connectedk8s/azext_connectedk8s/custom.py:761]: Yeah, that sounds good.

**Review by akashkeshari — APPROVED** [review]

**Comment by alphaWizard** [comment]
@fengzhou-msft review is done from the connectedk8s side. It is ready for checking in if you don't have any further reviews.

File: src/connectedk8s/setup.py [hunk 8]
```diff
@@ -17,7 +17,7 @@
 # TODO: Confirm this is the right version number you want and it matches your
 # HISTORY.rst entry.
 
-VERSION = '0.2.7'
+VERSION = '0.2.8'
```
Comment (Anumita) [inline src/connectedk8s/setup.py:20]: revert this change. We will be raising another PR with the increase in version

**Review by Anumita — CHANGES_REQUESTED** [review]

**Comment by Anumita** [comment]
> @fengzhou-msft review is done from the connectedk8s side. It is ready for checking in if you don't have any further reviews.

hey @fengzhou-msft , will ping for merging once @alphaWizard have reverted the setup.py change

File: src/connectedk8s/setup.py (hunk 8 above)
Comment (alphaWizard) [reply src/connectedk8s/setup.py]: Done

**Review by Anumita — APPROVED** [review]

**Comment by alphaWizard** [comment]
@fengzhou-msft Review is done from the connectedk8s side. It is ready for checking in if you don't have any further reviews.

**Comment by alphaWizard** [comment]
> @fengzhou-msft Review is done from the connectedk8s side. It is ready for checking in if you don't have any further reviews.

@fengzhou-msft Gentle ping regarding this.

**Review by fengzhou-msft — APPROVED** [review]
